---
TITLE: "GLM for transcription elongation"
AUTHORS: "Lingjie Liu; Yixin Zhao; Rebecca Hasset; Adam Siepel"
---

# Probabilistic and machine-learning methods for predicting local rates of transcription elongation from nascent RNA sequencing data

## Overview
In this study, we revisit the questions about the determinants of local elongation rates through gene bodies using a fundamentally different statistical modeling approach.  Our method is based on a recently developed "unified model" for nascent RNA sequencing (NRS) data, which describes both the kinetics of Pol II movement on the DNA template and the generation of NRS read counts (see [Siepel 2022](https://www.biorxiv.org/content/10.1101/2021.01.12.426408v1) and [Zhao et al 2023](https://academic.oup.com/nar/article/51/21/e106/7331020)). 

We adapt this model to allow for a continuously variable elongation rate along the genome, using a generalized linear model to capture the relationship between the local rate and nearby genomic features (see [Liu et al 2023](https://www.biorxiv.org/content/10.1101/2023.12.21.572932v1)). By accounting for differences across genes in initiation rate, we are able to efficiently pool information across genes, and extract high-resolution information about relative local elongation rates from steady-state data. 

Here, we provide scripts of the implementation, and demonstrate how quantify both previously known and novel correlates of elongation rate through estimation of coefficients $\kappa$. We then use our models to predict nucleotide-specific elongation rates $\zeta_i$ genome-wide and make our predictions available in a UCSC Genome Browser track.

<p align="center">
  <img src="figures/concept.png" alt="glm" width="1000"/>
</p>

<p align = "left">
	Fig. 1 Conceptual illustration of kinetic model for Pol II movement along DNA template in gene body. At nucleotide site $i$, relative elongation rate $\zeta_i$ is an exponentiated linear function of features $\vec Y_i$ and coefficients $\kappa$. 
	Promoter-proximal pausing and termination are ignored here. 
	Graphical model representation showing unobserved continuous-time Markov chain ($Z_i$) and observed NRS read counts ($X_i$).
	Conceptual illustration showing that differences in average gene-body read depth are explained by the scaled initiation rate $\chi$, while relative read depth is explained by the generalized linear model for local elongation rate $\zeta_i$. 
	Read count $X_i$ is assumed to be Poisson distributed with mean $\frac{\chi}{\zeta_i}$.
</p>

## Dependencies

GLM is implemented in the statistical programming language [R](https://www.r-project.org/), and depends on a couple of packages. One of the easiest ways to install them is via [conda](https://docs.conda.io/en/latest/).

```
conda env create --file pub_glm_env.yaml
```

Once installed, you can activate the environment then run the examples within it,

```
conda activate pub_glm
```

Data for the examples below can be downloaded from [Galaxy](https://usegalaxy.org/u/lingjie_liu/h/glm-test-data), and should be assumed to be stored within the data directory. 


## Examples

### Estimate coefficients $\kappa$ based on epigenomic model

```
usage: Rscript ./estimate_epigenomic_kappa.R [-h] -i InputFile [-c Category]
                                             [-l_s LearningSize]
                                             [-t Tolerance] [-o OutputDir]

Estimate coefficients kappa of epigenomic model

options:
  -h, --help         show this help message and exit
  -i InputFile       input training set of gene coordination, read counts and
                     covariates
  -c Category        category of input (name of cell line or simulation)
                     [default "simulation"]
  -l_s LearningSize  learning size for gradient ascent [default 1e-07]
  -t Tolerance       tolerance for gradient ascent [default 0.1]
  -o OutputDir       directory for saving the output of kappa
```

1. To validate the our framework of 
The input data is produced by [SimPol](https://github.com/CshlSiepelLab/SimPol), a simulator we developed for simulating the dynamics of RNA Pol II on DNA template. 
A .csv file is generated by SimPol for each simulated transcription unit (TU). 100 TUs are simulated and 80% serves as training data.

```
Rscript ./estimate_epigenomic_kappa.R -i ../data/simEP_gbrcGaussian_trainOne.Rdata -c simulation -l_s 1e-6 -t 1e-2 
```

2. Work with real data.
```
Rscript ./estimate_epigenomic_kappa.R -i ../data/k562_samp_epft_norm_train_1.Rdata -c k562 -l_s 1e-7 -t 1e-1
```


### Estimate coefficients $\kappa$ based on k-mer model ($k \in \{1, 2, 3, 4, 5\}$)

```
usage: Rscript ./estimate_kmer_kappa.R [-h] -i InputFile -k_m Inputkmer -k_t
                                       InputkmerTypes [-c Category]
                                       [-l_s LearningSize] [-t Tolerance]
                                       [-p penalty] [-o OutputDir]

Estimate coefficients kappa of k-mer model

options:
  -h, --help           show this help message and exit
  -i InputFile         input training set of gene coordination, read counts
  -k_m Inputkmer       input training set of k-mer sparse matrix
  -k_t InputkmerTypes  types for all kmers (AATTA, AATTT...)
  -c Category          category of input (name of cell line or simulation)
                       [default "simulation"]
  -l_s LearningSize    learning size for gradient ascent [default 1e-07]
  -t Tolerance         tolerance for gradient ascent [default 0.01]
  -p penalty           hyperparameter for L1 penalty, log10(nu) [default -6]
  -o OutputDir         directory for saving the output of kappa
```

1. Work with simulated data.
```
Rscript ./estimate_kmer_kappa.R -i ../data/simKmer_gbrc_trainAll.Rdata -k_m ../data/simKmer_kmerMT_trainAll.Rdata -k_t ../data/allmer_types.RData -c simulation -l_s 1e-4 -t 1e-2 -p -5.8
```

2. Work with real data.
```
Rscript ./estimate_kmer_kappa.R -i ../data/k562_subset1_gbrc_train.RData -k_m ../data/k562_subset1_allmerMT_train.RData -k_t ../data/allmer_types.RData -c k562 -l_s 1e-7 -t 1e-1 -p -3.6
```


### Predicted nucleotide-specific elongation rates $\zeta_i$

```
usage: Rscript ./predict_zeta.R [-h] -g GeneBody [-c Category] -r ReadCount
                                [-m Model] -e_m EpMatrix -k Kappa
                                [-k_m KmerMatrix] [-o OutputDir]

Predict nucleotide-specific elongation rate zeta

options:
  -h, --help       show this help message and exit
  -g GeneBody      input of gene coordinates in the predicted regions
  -c Category      category of input (name of cell line) [default "k562"]
  -r ReadCount     input of PRO-seq read counts in the predicted regions
  -m Model         select a model (epigenomic, combined) [default
                   "epigenomic"]
  -e_m EpMatrix    epigenomic covariates in the predicted regions
  -k Kappa         estimated kappa values for epigenomic or combined models
  -k_m KmerMatrix  input of k-mer sparse matrix in the predicted regions
  -o OutputDir     directory for saving the predicted nucleotide-specific zeta
```

1. Predict using the epigenomic model.
```
Rscript ./predict_zeta.R -g ../data/shared_gb_twoGenes.Rdata -c k562 -r ../data/k562_rc_twoGenes.Rdata -m epigenomic -e_m ../data/k562_epft_norm_twoGenes.Rdata -k ../data/epKappa_fourCell.csv
```

2.  Predict using the combined model (epigenomic and k-mer features).
```
Rscript ./predict_zeta.R -g ../data/shared_gb_twoGenes.Rdata -c k562 -r ../data/k562_rc_twoGenes.Rdata -m combined -e_m ../data/k562_epft_norm_twoGenes.Rdata -k ../data/epAllmerKappa_fourCell.csv -k_m ../data/allmerMT_twoGenes.Rdata
```



## Citation
https://www.biorxiv.org/content/10.1101/2021.01.12.426408v1
