---
TITLE: "GLM for transcription elongation"
AUTHORS: "Lingjie Liu; Yixin Zhao; Rebecca Hasset; Adam Siepel"
---

# Probabilistic and machine-learning methods for predicting local rates of transcription elongation from nascent RNA sequencing data

## Overview
In our manuscript, we devise a probabilistic model that predicts nucleotide-specific elongation rates as a generalized linear function of nearby genomic and epigenomic features. 
Here, we provide scripts of the implementation, and demonstrate how we can use them to estimate coefficient $\kappa$ for both synthetic and experimental data.

<p align="center">
  <img src="figures/concept.png" alt="glm" width="1000"/>
</p>

<p align = "left">
	Fig. 1 Conceptual illustration of kinetic model for Pol II movement along DNA template in gene body. At nucleotide site $i$, relative elongation rate $\zeta_i$ is an exponentiated linear function of features $\vec Y_i$ and coefficients $\kappa$. Promoter-proximal pausing and termination are ignored here. 
	Graphical model representation showing unobserved continuous-time Markov chain ($Z_i$) and observed NRS read counts ($X_i$).
	Conceptual illustration showing that differences in average gene-body read depth are explained by the scaled initiation rate $\chi$, while relative read depth is explained by the generalized linear model for local elongation rate $\zeta_i$. 
	Read count $X_i$ is assumed to be Poisson distributed with mean $\frac{\chi}{\zeta_i}$.
</p>

## Dependencies

GLM is implemented in the statistical programming language [R](https://www.r-project.org/), and depends on a couple of packages. One of the easiest ways to install them is via [conda](https://docs.conda.io/en/latest/).

```
conda create -n pub_glm --file pub_glm_env.yaml
```

Once installed, you can activate the environment then run the examples within it,

```
conda activate pub_glm
```

Data for the examples below can be downloaded from [Galaxy](https://usegalaxy.org/u/lingjie_liu/h/glm-test-data), and should be assumed to be stored within the data directory. 


## Examples

### Estimate coefficients $\kappa$ based on simulated data

```
usage: ./estimate_epigenomic_kappa.R [-h] -i InputFile [-c Category]
                                     [-l_s LearningSize] [-t Tolerance]
                                     [-o OutputDir]

Estimate coefficients kappa of epigenomic model

options:
  -h, --help         show this help message and exit
  -i InputFile       input training set of gene coordination, read counts and
                     covariates
  -c Category        category of input (name of cell line or simulation)
                     [default "simulation"]
  -l_s LearningSize  learning size for gradient ascent [default 1e-07]
  -t Tolerance       tolerance for gradient ascent [default 0.1]
  -o OutputDir       directory for saving the output of kappa
```

The input data is produced by [SimPol](https://github.com/CshlSiepelLab/SimPol), a simulator we developed for simulating the dynamics of RNA Pol II on DNA template. 
A .csv file is generated by SimPol for each simulated transcription unit (TU). 100 TUs are simulated and 80% serves as training data.

```
Rscript ./estimate_epigenomic_kappa.R -i ../data/simEP_gbrcGaussian_trainOne.Rdata -c simulation -l_s 1e-6 -t 1e-2 
```
